spring-cloud-security/src/main/java/org/springframework/cloud/security/oauth2/client/feign/OAuth2FeignRequestInterceptor.java

```diff
@@ -0,0 +1,107 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.security.oauth2.client.feign;
+
+import feign.RequestInterceptor;
+import feign.RequestTemplate;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.security.oauth2.client.OAuth2ClientContext;
+import org.springframework.security.oauth2.client.token.AccessTokenRequest;
+
+/**
+ * Pre-defined custom RequestInterceptor for Feign Requests
+ * It uses the {@link OAuth2ClientContext OAuth2ClientContext} provided from the environment
+ * and construct a new header on the request before it is made by Feign
+ *
+ * @author Joao Pedro Evangelista
+ */
+public class OAuth2FeignRequestInterceptor implements RequestInterceptor {
+
+    public static final String BEARER = "Bearer";
+
+    public static final String AUTHORIZATION = "Authorization";
+
+    private final Logger logger = LoggerFactory.getLogger(OAuth2FeignRequestInterceptor.class);
+
+    private final OAuth2ClientContext oAuth2ClientContext;
+
+    private final String tokenType;
+
+    private final String header;
+
+    /**
+     * Default constructor which uses the provided OAuth2ClientContext and
+     * Bearer tokens within Authorization header
+     *
+     * @param oAuth2ClientContext provided context
+     */
+    public OAuth2FeignRequestInterceptor(OAuth2ClientContext oAuth2ClientContext) {
+        this(oAuth2ClientContext, BEARER, AUTHORIZATION);
+        logger.debug("Constructing default OAuth2FeignRequestInterceptor");
+    }
+
+    /**
+     * Fully customizable constructor for changing token type and header name, in cases of Bearer and Authorization is not the default
+     * such as "bearer", "authorization"
+     *
+     * @param oAuth2ClientContext current oAuth2 Context
+     * @param tokenType           type of token e.g. "token", "Bearer"
+     * @param header              name of the header e.g. "Authorization", "authorization"
+     */
+    public OAuth2FeignRequestInterceptor(OAuth2ClientContext oAuth2ClientContext, String tokenType, String header) {
+        this.oAuth2ClientContext = oAuth2ClientContext;
+        this.tokenType = tokenType;
+        this.header = header;
+    }
+
+    private static boolean tokenExists(OAuth2ClientContext oAuth2ClientContext) {
+        return oAuth2ClientContext.getAccessTokenRequest().getExistingToken() != null;
+    }
+
+    /**
+     * Create a template with the header of provided name and extracted value
+     *
+     * @see RequestInterceptor#apply(RequestTemplate)
+     */
+    @Override
+    public void apply(RequestTemplate template) {
+        if (tokenExists(oAuth2ClientContext)) {
```

Comment:
You could probably make this check optional. That's the way the `OAuth2RestTemplate` is usually used - a `Filter` catches the exception and redirects to the auth server.